server/tserver/src/main/java/org/apache/accumulo/tserver/Tablet.java

```diff
@@ -1761,33 +1763,41 @@ Scanner createScanner(Range range, int num, Set<Column> columns, Authorizations
     private ScanDataSource isolatedDataSource;
     private boolean sawException = false;
     private boolean scanClosed = false;
+    private Semaphore scannerSemaphore;
 
     Scanner(Range range, ScanOptions options) {
       this.range = range;
       this.options = options;
+      scannerSemaphore = new Semaphore(1, true);
     }
 
-    synchronized ScanBatch read() throws IOException, TabletClosedException {
+    ScanBatch read() throws IOException, TabletClosedException {
 
-      if (sawException)
-        throw new IllegalStateException("Tried to use scanner after exception occurred.");
-
-      if (scanClosed)
-        throw new IllegalStateException("Tried to use scanner after it was closed.");
+      ScanDataSource dataSource = null;
 
       Batch results = null;
 
-      ScanDataSource dataSource;
+      try {
 
-      if (options.isolated) {
-        if (isolatedDataSource == null)
-          isolatedDataSource = new ScanDataSource(options);
-        dataSource = isolatedDataSource;
-      } else {
-        dataSource = new ScanDataSource(options);
-      }
+        try {
+          scannerSemaphore.acquire();
+        } catch (InterruptedException e) {
+          sawException = true;
+        }
 
-      try {
+        if (sawException)
+          throw new IllegalStateException("Tried to use scanner after exception occurred.");
```

Comment:
Isn't this exception message a little misleading now if we catch the InterruptedException above? Would it be better to throw an ISE from the catch block above with a better message? Or, do we not care what exception we saw?